src/cman.jl

```diff
@@ -59,268 +92,392 @@ type MPIManager <: ClusterManager
-         end
-     elseif mode == MPI_TRANSPORT_ALL
+        end
+    elseif mode == MPI_TRANSPORT_ALL
+        comm = MPI.Comm_dup(comm)
+        rank = MPI.Comm_rank(comm)
+        size = MPI.Comm_size(comm)
+        # We are handling the workers and their event loops on our own
         if rank == 0
-            cman = MPIManager(;np=comm_size-1, mode=MPI_TRANSPORT_ALL)
-            @schedule main_loop(cman, comm)
-            addprocs(cman)
-            return cman
+            # On the manager:
+            # Create manager object
+            mgr = MPIManager(np=size-1, mode=mode)
+            mgr.comm = comm
+            # Start event loop for the workers
+            @schedule receive_event_loop(mgr)
+            # Tell Base about the workers
+            addprocs(mgr)
+            return mgr
         else
-            cman = MPIManager(;np=comm_size-1, mode=MPI_TRANSPORT_ALL)
-            Base.init_worker(cman)
-            main_loop(cman, comm) # does not return
+            # On a worker:
+            # Create a "fake" manager object since Base wants one
+            mgr = MPIManager(np=size-1, mode=mode)
+            mgr.comm = comm
+            Base.init_worker(mgr)
+            # Start a worker event loop
+            receive_event_loop(mgr)
+            MPI.Finalize()
+            exit()
         end
-     else
-        error("Unknown Mode $mode")
-     end
+    else
+        error("Unknown mode $mode")
+    end
 end
 
-function main_loop(manager, comm)
-    while !MPI.Finalized()
-        (hasdata, stat) = MPI.Iprobe(ANY_SOURCE, 0, comm)
+# Event loop for receiving data, for the MPI_TRANSPORT_ALL case
+function receive_event_loop(mgr::MPIManager)
+    num_send_loops = 0
+    while !isready(mgr.initiate_shutdown)
+        (hasdata, stat) = MPI.Iprobe(MPI.ANY_SOURCE, 0, mgr.comm)
         if hasdata
             count = Get_count(stat, UInt8)
             buf = Array(UInt8, count)
             from_rank = Get_source(stat)
-            MPI.Recv!(buf, from_rank, 0, comm)
+            MPI.Recv!(buf, from_rank, 0, mgr.comm)
 
-            streams = get(manager.id_to_streams, from_rank, nothing)
+            streams = get(mgr.rank2streams, from_rank, nothing)
             if streams == nothing
-                # First time..
-                (r_s, w_s) = setup_connection(manager, from_rank)
+                # This is the first time we communicate with this rank.
+                # Set up a new connection.
+                (r_s, w_s) = start_send_event_loop(mgr, from_rank)
                 Base.process_messages(r_s, w_s)
+                num_send_loops += 1
             else
                 (r_s, w_s) = streams
             end
             write(r_s, buf)
         else
-            # FIXME : Need a better way to integrate with libuv's event loop
+            # TODO: Need a better way to integrate with libuv's event loop
             yield()
         end
     end
+    for i in 1:num_send_loops
+        fetch(mgr.sending_done)
+    end
```

Comment:
@eschnett How are `num_send_loops`, `sending_done` and `receiving_done` being used? The `fetch` is not a destructive read - it will just return the first value `num_send_loops` times.